File: backend/compras/src/model/compra_pedido_detalhe_model.py
from src import db
from src.model.produto_model import ProdutoModel


class CompraPedidoDetalheModel(db.Model):
    __tablename__ = 'compra_pedido_detalhe'

    id = db.Column(db.Integer, primary_key=True)
    quantidade = db.Column(db.Float)
    valor_unitario = db.Column(db.Float)
    valor_subtotal = db.Column(db.Float)
    taxa_desconto = db.Column(db.Float)
    valor_desconto = db.Column(db.Float)
    valor_total = db.Column(db.Float)
    cst = db.Column(db.String(2))
    csosn = db.Column(db.String(3))
    cfop = db.Column(db.Integer)
    base_calculo_icms = db.Column(db.Float)
    valor_icms = db.Column(db.Float)
    valor_ipi = db.Column(db.Float)
    aliquota_icms = db.Column(db.Float)
    aliquota_ipi = db.Column(db.Float)
    id_compra_pedido = db.Column(db.Integer, db.ForeignKey('compra_pedido.id'))
    id_produto = db.Column(db.Integer, db.ForeignKey('produto.id'))

    produto_model = db.relationship('ProdutoModel', foreign_keys=[id_produto])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_compra_pedido = data.get('idCompraPedido')
        self.id_produto = data.get('idProduto')
        self.quantidade = data.get('quantidade')
        self.valor_unitario = data.get('valorUnitario')
        self.valor_subtotal = data.get('valorSubtotal')
        self.taxa_desconto = data.get('taxaDesconto')
        self.valor_desconto = data.get('valorDesconto')
        self.valor_total = data.get('valorTotal')
        self.cst = data.get('cst')
        self.csosn = data.get('csosn')
        self.cfop = data.get('cfop')
        self.base_calculo_icms = data.get('baseCalculoIcms')
        self.valor_icms = data.get('valorIcms')
        self.valor_ipi = data.get('valorIpi')
        self.aliquota_icms = data.get('aliquotaIcms')
        self.aliquota_ipi = data.get('aliquotaIpi')

    def serialize(self):
        return {
            'id': self.id,
            'idCompraPedido': self.id_compra_pedido,
            'idProduto': self.id_produto,
            'quantidade': self.quantidade,
            'valorUnitario': self.valor_unitario,
            'valorSubtotal': self.valor_subtotal,
            'taxaDesconto': self.taxa_desconto,
            'valorDesconto': self.valor_desconto,
            'valorTotal': self.valor_total,
            'cst': self.cst,
            'csosn': self.csosn,
            'cfop': self.cfop,
            'baseCalculoIcms': self.base_calculo_icms,
            'valorIcms': self.valor_icms,
            'valorIpi': self.valor_ipi,
            'aliquotaIcms': self.aliquota_icms,
            'aliquotaIpi': self.aliquota_ipi,
            'produtoModel': self.produto_model.serialize() if self.produto_model else None,
        }

File: frontend/compras/lib/app/data/model/compra_pedido_detalhe_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:compras/app/data/model/model_imports.dart';

class CompraPedidoDetalheModel {
	int? id;
	int? idCompraPedido;
	int? idProduto;
	double? quantidade;
	double? valorUnitario;
	double? valorSubtotal;
	double? taxaDesconto;
	double? valorDesconto;
	double? valorTotal;
	String? cst;
	String? csosn;
	int? cfop;
	double? baseCalculoIcms;
	double? valorIcms;
	double? valorIpi;
	double? aliquotaIcms;
	double? aliquotaIpi;
	ProdutoModel? produtoModel;

	CompraPedidoDetalheModel({
		this.id,
		this.idCompraPedido,
		this.idProduto,
		this.quantidade,
		this.valorUnitario,
		this.valorSubtotal,
		this.taxaDesconto,
		this.valorDesconto,
		this.valorTotal,
		this.cst,
		this.csosn,
		this.cfop,
		this.baseCalculoIcms,
		this.valorIcms,
		this.valorIpi,
		this.aliquotaIcms,
		this.aliquotaIpi,
		this.produtoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'quantidade',
		'valor_unitario',
		'valor_subtotal',
		'taxa_desconto',
		'valor_desconto',
		'valor_total',
		'cst',
		'csosn',
		'cfop',
		'base_calculo_icms',
		'valor_icms',
		'valor_ipi',
		'aliquota_icms',
		'aliquota_ipi',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Quantidade',
		'Valor Unitario',
		'Valor Subtotal',
		'Taxa Desconto',
		'Valor Desconto',
		'Valor Total',
		'Cst',
		'Csosn',
		'Cfop',
		'Base Calculo Icms',
		'Valor Icms',
		'Valor Ipi',
		'Aliquota Icms',
		'Aliquota Ipi',
	];

	CompraPedidoDetalheModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCompraPedido = jsonData['idCompraPedido'];
		idProduto = jsonData['idProduto'];
		quantidade = jsonData['quantidade']?.toDouble();
		valorUnitario = jsonData['valorUnitario']?.toDouble();
		valorSubtotal = jsonData['valorSubtotal']?.toDouble();
		taxaDesconto = jsonData['taxaDesconto']?.toDouble();
		valorDesconto = jsonData['valorDesconto']?.toDouble();
		valorTotal = jsonData['valorTotal']?.toDouble();
		cst = jsonData['cst'];
		csosn = jsonData['csosn'];
		cfop = jsonData['cfop'];
		baseCalculoIcms = jsonData['baseCalculoIcms']?.toDouble();
		valorIcms = jsonData['valorIcms']?.toDouble();
		valorIpi = jsonData['valorIpi']?.toDouble();
		aliquotaIcms = jsonData['aliquotaIcms']?.toDouble();
		aliquotaIpi = jsonData['aliquotaIpi']?.toDouble();
		produtoModel = jsonData['produtoModel'] == null ? ProdutoModel() : ProdutoModel.fromJson(jsonData['produtoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCompraPedido'] = idCompraPedido != 0 ? idCompraPedido : null;
		jsonData['idProduto'] = idProduto != 0 ? idProduto : null;
		jsonData['quantidade'] = quantidade;
		jsonData['valorUnitario'] = valorUnitario;
		jsonData['valorSubtotal'] = valorSubtotal;
		jsonData['taxaDesconto'] = taxaDesconto;
		jsonData['valorDesconto'] = valorDesconto;
		jsonData['valorTotal'] = valorTotal;
		jsonData['cst'] = cst;
		jsonData['csosn'] = csosn;
		jsonData['cfop'] = cfop;
		jsonData['baseCalculoIcms'] = baseCalculoIcms;
		jsonData['valorIcms'] = valorIcms;
		jsonData['valorIpi'] = valorIpi;
		jsonData['aliquotaIcms'] = aliquotaIcms;
		jsonData['aliquotaIpi'] = aliquotaIpi;
		jsonData['produtoModel'] = produtoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCompraPedido = plutoRow.cells['idCompraPedido']?.value;
		idProduto = plutoRow.cells['idProduto']?.value;
		quantidade = plutoRow.cells['quantidade']?.value?.toDouble();
		valorUnitario = plutoRow.cells['valorUnitario']?.value?.toDouble();
		valorSubtotal = plutoRow.cells['valorSubtotal']?.value?.toDouble();
		taxaDesconto = plutoRow.cells['taxaDesconto']?.value?.toDouble();
		valorDesconto = plutoRow.cells['valorDesconto']?.value?.toDouble();
		valorTotal = plutoRow.cells['valorTotal']?.value?.toDouble();
		cst = plutoRow.cells['cst']?.value;
		csosn = plutoRow.cells['csosn']?.value;
		cfop = plutoRow.cells['cfop']?.value;
		baseCalculoIcms = plutoRow.cells['baseCalculoIcms']?.value?.toDouble();
		valorIcms = plutoRow.cells['valorIcms']?.value?.toDouble();
		valorIpi = plutoRow.cells['valorIpi']?.value?.toDouble();
		aliquotaIcms = plutoRow.cells['aliquotaIcms']?.value?.toDouble();
		aliquotaIpi = plutoRow.cells['aliquotaIpi']?.value?.toDouble();
		produtoModel = ProdutoModel();
		produtoModel?.nome = plutoRow.cells['produtoModel']?.value;
	}	

	CompraPedidoDetalheModel clone() {
		return CompraPedidoDetalheModel(
			id: id,
			idCompraPedido: idCompraPedido,
			idProduto: idProduto,
			quantidade: quantidade,
			valorUnitario: valorUnitario,
			valorSubtotal: valorSubtotal,
			taxaDesconto: taxaDesconto,
			valorDesconto: valorDesconto,
			valorTotal: valorTotal,
			cst: cst,
			csosn: csosn,
			cfop: cfop,
			baseCalculoIcms: baseCalculoIcms,
			valorIcms: valorIcms,
			valorIpi: valorIpi,
			aliquotaIcms: aliquotaIcms,
			aliquotaIpi: aliquotaIpi,
		);			
	}

	
}